[FLINK-25430] Replace RunningJobRegistry by JobResultStore

XComp opened the pull request:
## What is the purpose of the change

This PR introduces the `JobResultStore` interface that replaces the `RunningJobRegistry` previously used in the `Dispatcher`.

The functionality changes in a way that the no persisting registry for running jobs is being provided by the HA implementations (k8s and ZK). This will be reintroduced by the changes made as part of [FLINK-25431](https://issues.apache.org/jira/browse/FLINK-25431)

## Brief change log

* Introduced `JobResultStore`
* Implemented `EmbeddedJobResultStore`
* Removed `RunningJobRegistry` interface and its implementing components
* Integrates `JobResultStore` into `Dispatcher`

## Verifying this change

* Available tests should work as usual since no functionality changed
* `EmbeddedJobResultStoreTest` implemented

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no (this change is only a refactoring)
  - If yes, how is the feature documented? not applicable

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunnerITCase.java:
```diff
@@ -262,7 +272,17 @@ private DispatcherRunner createDispatcherRunner() throws Exception {
                 dispatcherLeaderElectionService,
                 fatalErrorHandler,
-                () -> jobGraphStore,
+                new JobPersistenceComponentFactory() {
```
Should we have a testing implementation that allows you to pass the stores via constructor instead? (similar to `TestingCheckpointRecoveryFactory`)

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -160,7 +168,7 @@ public void closeAsync_stopsJobGraphStoreAndDispatcher() throws Exception {
                 TestingDispatcherServiceFactory.newBuilder()
                         .setCreateFunction(
-                                (ignoredA, ignoredB, ignoredC) ->
+                                (ignoredA, ignoredB, ignoredC, ignoredD, ignoredE) ->
```
see comments on `QuintFunction`

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/ZooKeeperDefaultDispatcherRunnerTest.java:
```diff
@@ -180,7 +179,17 @@ public void testResourceCleanupUnderLeadershipChange() throws Exception {
                             rpcService,
                             dispatcherLeaderElectionService,
-                            () -> createZooKeeperJobGraphStore(client),
+                            new JobPersistenceComponentFactory() {
```
again, `new TestingJobPersistenceComponentFactory(createZooKeeperJobGraphStore(client), new EmbeddedJobResultStore())` might be nice to have

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperHaServicesTest.java:
```diff
@@ -235,8 +235,7 @@ private void runCleanupTestWithJob(
 
-            final RunningJobsRegistry runningJobsRegistry =
-                    zooKeeperHaServices.getRunningJobsRegistry();
+            final JobResultStore jobResultStore = zooKeeperHaServices.getJobResultStore();
```
unused, should we also incorporate dirty -> clean into the lifecycle?

XComp replied:
good catch. We should remove it. The test is about cleaning up the znodes. It's just about creating some paths. The integration of this component into the test becomes obsolete with the `JobResultStore` replacing the `RunningJobRegistry` since no ZK-related artifacts are created by the `JobResultStore`.

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunnerTest.java:
```diff
@@ -662,15 +663,21 @@ public void testJobMasterServiceProcessCreationFailureIsForwardedToResultFuture(
     @Test
     public void testJobAlreadyDone() throws Exception {
-        JobID jobID = new JobID();
+        JobID jobId = new JobID();
+        JobResult jobResult =
```
nit
```suggestion
        final JobID jobId = new JobID();
        final JobResult jobResult =
```

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStoreTest.java:
```diff
@@ -0,0 +1,83 @@
+import static org.junit.Assert.assertTrue;
+
+/** Tests for the {@link EmbeddedJobResultStore}. */
+public class EmbeddedJobResultStoreTest extends TestLogger {
```
nit: junit5 + assertj

dmvk reviewed (changes requested):
Great job @autophagy and @XComp, thanks for the PR! This already looks really solid, I've added few comments, PTAL. My biggest concern would be around thread safety of `(Embedded)?JobResultStore` and the JRS usage in combination with the per job mode (MiniDispatcher).

Maybe since I did a full pass on the PR, would it make sense to squash the commit history before addressing the comments?

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -868,6 +867,14 @@ private void cleanUpRemainingJobData(JobID jobId, boolean jobGraphRemoved) {
+        try {
+            jobResultStore.markResultAsClean(jobId);
+        } catch (IOException e) {
+            log.warn("Could not properly mark job {} result as clean.", jobId, e);
```
Yes, this would be addressed when introducing the retry mechanism

dmvk replied:
Maybe should we just re-throw a runtime exception here for now? So we don't forget to fix it.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcess.java:
```diff
@@ -92,34 +99,61 @@ private void startServices() {
-        return CompletableFuture.supplyAsync(this::recoverJobsIfRunning, ioExecutor);
+    private CompletableFuture<Void>
+            createDispatcherBasedOnRecoveredJobGraphsAndGloballyTerminatedJobs() {
+        return CompletableFuture.supplyAsync(
+                        this::getGloballyCompletedJobResultsIfRunning, ioExecutor)
+                .thenCompose(
+                        globallyTerminatedJobs ->
+                                CompletableFuture.supplyAsync(
+                                                () ->
+                                                        this.recoverJobsIfRunning(
+                                                                globallyTerminatedJobs.stream()
+                                                                        .map(JobResult::getJobId)
+                                                                        .collect(
+                                                                                Collectors
+                                                                                        .toSet())),
+                                                ioExecutor)
+                                        .thenAccept(
+                                                jobGraphs ->
+                                                        createDispatcherIfRunning(
+                                                                jobGraphs, globallyTerminatedJobs))
+                                        .handle(this::onErrorIfRunning));
```
Interesting, I didn't know about the `.thenAcceptBoth` method. That makes the code way easier to read 👍

XComp replied:
I'm wondering whether the `dirtyJobsFuture.thenApplyAsync` is necessary here. Wouldn't it be executed in the same thread pool as the previous `supplyAsync` since both are chained together?

dmvk replied:
I've tried to reply via email, but it doesn't seem to have arrived :/

It would be executed in the same thread that has called the complete method (which belongs to the pool). Passing the executor directly allows another thread from the pool to execute the callback. Also the control mechanism is kind of more explicit as you can never be sure that the "complete" method has been called by the thread from previous stage.

dmvk commented:
It would be executed in the same thread that has called the complete method
(which belongs to the pool). Passing the executor directly allows for
picking another thread from the pool. Also the control mechanism it's kind
of more explicit as you can never be sure that the "complete" method has
been called by the thread from previous stage.

On Fri, Jan 7, 2022 at 4:04 PM Matthias Pohl ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcess.java
> <https://github.com/apache/flink/pull/18189#discussion_r780321496>:
>
> > +        return CompletableFuture.supplyAsync(
> +                        this::getGloballyCompletedJobResultsIfRunning, ioExecutor)
> +                .thenCompose(
> +                        globallyTerminatedJobs ->
> +                                CompletableFuture.supplyAsync(
> +                                                () ->
> +                                                        this.recoverJobsIfRunning(
> +                                                                globallyTerminatedJobs.stream()
> +                                                                        .map(JobResult::getJobId)
> +                                                                        .collect(
> +                                                                                Collectors
> +                                                                                        .toSet())),
> +                                                ioExecutor)
> +                                        .thenAccept(
> +                                                jobGraphs ->
> +                                                        createDispatcherIfRunning(
> +                                                                jobGraphs, globallyTerminatedJobs))
> +                                        .handle(this::onErrorIfRunning));
>
> I'm wondering whether the dirtyJobsFuture.thenApplyAsync is necessary
> here. Wouldn't it be executed in the same thread pool as the previous
> supplyAsync since both are chained together?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink/pull/18189#discussion_r780321496>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACJGBL5ED6XK77BRIONM4DUU36GPANCNFSM5KUWUGAA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java:
```diff
@@ -130,11 +130,11 @@ public JobGraphStore getJobGraphStore() throws Exception {
-        if (runningJobsRegistry == null) {
-            this.runningJobsRegistry = createRunningJobsRegistry();
+    public JobResultStore getJobResultStore() throws Exception {
+        if (jobResultStore == null) {
```
Good point, that's some leftover from previous `RunningJobsRegistry` where I didn't focus enough on cleaning it up

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/AbstractHaServices.java:
```diff
@@ -244,11 +244,13 @@ public void cleanupJobData(JobID jobID) throws Exception {
      */
-    protected abstract RunningJobsRegistry createRunningJobsRegistry();
+    protected JobResultStore createJobResultStore() {
+        return new EmbeddedJobResultStore();
```
I removed this method entirely. The `JobResultStore` is passed in as a parameter through the constructor now. The actual implementations of `AbstractHaServices` are forced to instantiate it and pass it through the parent class constructor.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobPersistenceComponentFactory.java:
```diff
@@ -18,13 +18,22 @@
+    /**
+     * Creates {@link JobResultStore} instances.
+     *
+     * @return {@code JobResultStore} instances.
```
I prefer having only one link per JavaDoc paragraph (like it's done in Wikipedia) to reduce the distraction caused by link formatings when reading it. I aligned the JavaDoc of both methods in `JobPersistenceComponentFactory` accordingly. 👍

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/TestingDispatcher.java:
```diff
@@ -65,7 +68,14 @@ public void onStart() throws Exception {
+                    try {
+                        jobReachedTerminalState(executionGraphInfo);
+                    } catch (Exception e) {
+                        e.printStackTrace();
```
yikes, that one I overlooked 👍

XComp commented on flink-core/src/main/java/org/apache/flink/util/function/QuintFunction.java:
```diff
@@ -0,0 +1,48 @@
+ */
+@PublicEvolving
+@FunctionalInterface
+public interface QuintFunction<S, T, U, V, W, R> {
```
Can you clarify what you mean by "with five parameters, implementations are just too complex without parameter names"? You have to specify the parameter list when applying this class in any way (i.e. when using the actual interface or when using this generic Function). 🤔 

But besides that, I agree that `TestingDispatcherGatewayServiceFactory` does not add any value: I'm going to remove the `QuintFunction*` interfaces again and use `AbstractDispatcherLeaderProcess.DispatcherGatewayServiceFactory` as is in `SessionDispatcherLeaderProcessTest`.

dmvk replied:
I mean that the auto-completion automatically names all the parameter names for you, so you don't really have to think about what each parameter means, which is way more convenient for the person implementing it.

eg.
```
(a, b, c, d, e) -> {
}
```
vs.
```
(fencingToken, recoveredJobs, dirtyResults, ..., ...) -> {
}
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -196,6 +198,8 @@ public Dispatcher(
 
         this.recoveredJobs = new HashSet<>(recoveredJobs);
 
+        this.globallyTerminatedJobs = new HashSet<>(globallyTerminatedJobs);
```
These collections are cleared after the corresponding `JobManagerRunner` is created. The motivation behind that was, I guess, to save memory which sounds reasonable to me. If that's the case, using immutable sets for the `recoveredJobs` isn't a good idea.

I treated the `globallyTerminatedJobs` in the same way for consistency reasons. We could switch to an immutable set for the `globallyTerminatedJobs` member. But I don't see us gaining much from it. I'd rather prefer having it done analogously, instead.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/MiniDispatcher.java:
```diff
@@ -63,6 +63,7 @@ public MiniDispatcher(
                 rpcService,
                 fencingToken,
                 Collections.singleton(jobGraph),
+                Collections.emptyList(),
```
The JRS is meant to be scoped per cluster (basically the full job identifier should be tuple of `(clusterId, jobId)`). Future extension would be for example persisting the information that the cluster should shut down (this is needed to make shutdown in multi JM setup bullet proof).

Also with the application mode, the jobIds could easily clash as the fixed jobId is used there.

XComp replied:
Ah true, I forgot about that requirement. Thanks for clarification. I kept the general implementation anyway. 👍

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcess.java:
```diff
@@ -49,7 +50,9 @@ protected void onStart() {
                         DispatcherId.fromUuid(getLeaderSessionId()),
                         Collections.singleton(jobGraph),
-                        ThrowingJobGraphWriter.INSTANCE);
+                        Collections.emptyList(),
```
Same issue in the MiniDispatcher case. I left it as future task and forgot about it. It's fixed now

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntry.java:
```diff
@@ -0,0 +1,62 @@
+public class JobResultEntry {
+
+    private final JobResult jobResult;
+    private JobResultState state;
```
We changed the way where the `JobResultState` is referenced and how it's handled. We decided to move it out of the `JobResultEntry`. Instead, the `JobResultStore` implementation needs to take care of the `JobResultState`. That way we have a clear separation of responsibilities. The `JobResultEntry` is created outside of the `JobResultStore` and the `JobResultStore`'s responsibility is to manage the state of the `JobResultEntry` in some way. This makes `JobResultEntry` immutable.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Override
+    public void markResultAsClean(JobID jobId) throws NoSuchElementException {
```
Correct. Another thing I missed in my initial efforts combining the two PRs into a single one. 😇  This is fixed now. The `EmbeddedJobResultStore` holds two collections now: One for the dirty results and one for the clean ones. This has the advantage that we hold a dedicated index on the dirty entries.

XComp commented:
FYI: I didn't squash the initial commits, yet. The commits can be reviewed starting from `386d700c` (the first commit having the `[review]` label in its commit message).

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -943,15 +945,26 @@ protected CleanupJobState jobReachedTerminalState(ExecutionGraphInfo executionGr
-                jobResultStore.createDirtyResult(
-                        JobResult.createFrom(executionGraphInfo.getArchivedExecutionGraph()));
+                if (jobResultStore.hasCleanJobResultEntry(jobId)) {
+                    log.warn(
```
Should this be logged as a warning (at all)? What does this warning mean from user perspective? Does the user need to take any action to address that?

XComp replied:
I added it because it indicates that something strange is going on in the `JobResultStore` considering that the user could modify files in the directory of the file-based implementation. In general, this warning should never pop up. But if somebody did something wrong with file names, Flink might store the wrong state under that job.

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcess.java:
```diff
@@ -34,14 +37,21 @@
             UUID leaderSessionId,
             DispatcherGatewayServiceFactory dispatcherGatewayServiceFactory,
             JobGraph jobGraph,
+            Collection<JobResult> recoveredDirtyJobResults,
```
Would not having a collection here make sense? We already do the same thing with the jobGraph (recoveredJobs collection in other implementations)
```suggestion
            @Nullable JobResult recoveredDirtyJobResult,
```

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java:
```diff
@@ -58,12 +62,26 @@ public DispatcherLeaderProcessFactory createFactory(
+        final JobResultStore jobResultStore = jobPersistenceComponentFactory.createJobResultStore();
+        final Collection<JobResult> recoveredDirtyJobResults;
+        try {
+            recoveredDirtyJobResults = jobResultStore.getDirtyResults();
```
should we check that there is at most once result?

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultEntry.java:
```diff
@@ -22,41 +22,18 @@
+ * information about a globally terminated job (e.g. {@link JobResult}).
  */
-public class JobResultEntry {
+public class JobResultEntry implements WithJobResult {
```
Why do we need an extra `WithJobResult` interface?

XComp replied:
Initially, I though that it would be helpful for the file-based implementation (where we want to add the serializer version to persisted record). But looks like inheritance is good enough in that case. The interface is of no value. I'm going to delete this commit.

XComp replied:
I reverted the corresponding commit...

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/nonha/embedded/EmbeddedJobResultStore.java:
```diff
@@ -20,65 +20,112 @@
-    private final Map<JobID, JobResultEntry> inMemoryMap = new ConcurrentHashMap<>();
+    private static final Logger LOG = LoggerFactory.getLogger(EmbeddedJobResultStore.class);
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
```
👍

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/runner/SessionDispatcherLeaderProcessTest.java:
```diff
@@ -130,28 +146,56 @@ public void start_triggersJobGraphRecoveryAndDispatcherServiceCreation() throws
-                    dispatcherLeaderProcess.getState(),
-                    is(SessionDispatcherLeaderProcess.State.RUNNING));
+            assertThat(recoveredJobGraphsFuture)
+                    .succeedsWithin(100, TimeUnit.MILLISECONDS)
```
OT: We should rethink using, the `succeedsWithin` with assertj, I've run into this multiple times already. We should introduce something along the lines of `FlinkAssertions.assertThatSucceeds(future)....` with infinite timeout.

XComp replied:
Why is that again? Was it about making the test timeout which would provide a stacktrace of the waiting threads? 🤔

dmvk replied:
Yes. In case of regular timeout we lose any context that might be used for identifying the root cause.

XComp replied:
thanks for the pointer - I forgot about that requirement in the last PRs -.-

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/JobResultStore.java:
```diff
@@ -22,53 +22,88 @@
+ *       further actions need to be applied.
+ * </ul>
  */
 public interface JobResultStore {
```
```suggestion
@Internal
public interface JobResultStore {
```

dmvk commented on flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherGatewayServiceFactory.java:
```diff
@@ -118,7 +120,12 @@ public ApplicationDispatcherGatewayServiceFactory(
 
-    private List<JobID> getRecoveredJobIds(final Collection<JobGraph> recoveredJobs) {
-        return recoveredJobs.stream().map(JobGraph::getJobID).collect(Collectors.toList());
+    private Set<JobID> getRecoveredJobIds(
+            final Collection<JobGraph> recoveredJobs,
+            final Collection<JobResult> recoveredDirtyJobs) {
+        return Stream.concat(
+                        recoveredJobs.stream().map(JobGraph::getJobID),
+                        recoveredDirtyJobs.stream().map(JobResult::getJobId))
+                .collect(Collectors.toSet());
```
This part I'm not sure about. Could you document why we need to pass dirty jobs into ApplicationDispatcherBootstrap? Why is it not enough to just pass them into the dispatcher.

(I'm not saying it's incorrect, just can't reason about it from top of my head)

dmvk reviewed (commented):
Thanks for the update @XComp. I've added few more questions, mostly cosmetic questions. LGTM overall, great job

XComp commented:
> Maybe should we just re-throw a runtime exception here for now? So we don't forget to fix it.

I left it like that because that reflects the current state of the codebase. The `JobResultStore` entries will be left in a dirty state in such a case, which is the desired behavior. The user has to clean it up in that case manually. In the old (`master`) implementation we don't fail if the cleanup fails, either, but just write a warning to the logs.

XComp commented:
I addressed your comments and rebased the branch. I kept the old commits as is. You can review anything starting from the commits being labeled as `[review #2]`.

dmvk reviewed (commented):
LGTM overall, I've added few minor notes. Can you please try to fix the commit history into some smaller logical chunks?

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/JobDispatcherFactory.java:
```diff
@@ -38,13 +40,21 @@ public MiniDispatcher createDispatcher(
+                    partialDispatcherServicesWithJobPersistenceComponents)
             throws Exception {
-        final JobGraph jobGraph = Iterables.getOnlyElement(recoveredJobs);
+        final Optional<JobGraph> recoveredJobGraph = CollectionUtil.atMostOnce(recoveredJobs);
+        final Optional<JobResult> recoveredDirtyJob =
+                CollectionUtil.atMostOnce(recoveredDirtyJobResults);
+
+        Preconditions.checkArgument(
+                recoveredJobGraph.isPresent() != recoveredDirtyJob.isPresent(),
+                "Either the JobGraph or the recovered JobResult needs to be specified.");
 
```
Oh, interesting. That one I overlooked. Thanks 👍

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java:
```diff
@@ -52,22 +69,70 @@ public DispatcherLeaderProcessFactory createFactory(
         }
 
+        final JobResultStore jobResultStore = jobPersistenceComponentFactory.createJobResultStore();
+        final Collection<JobResult> recoveredDirtyJobResults = getDirtyJobResults(jobResultStore);
```
yeah, but I leave it like that now considering that the user is able to modify the file-based JobResultStore... That way, we make it more robust for such cases.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/JobDispatcherLeaderProcessFactoryFactory.java:
```diff
@@ -52,22 +69,70 @@ public DispatcherLeaderProcessFactory createFactory(
+                actualDirtyJobResult = Optional.of(dirtyJobResult);
+            } else {
+                LOG.warn(
+                        "Unexpected dirty JobResultStore entry: Job '{}' is listed as dirty, isn't part of this single-job cluster, though.",
```
I would leave it like that to have it being more robust to modifications done by the user in a file-based setup...

XComp commented on flink-clients/src/test/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrapITCase.java:
```diff
@@ -147,6 +157,67 @@ public void testDispatcherRecoversAfterLosingAndRegainingLeadership() throws Exc
+            awaitClusterStopped(cluster, deadline);
+        }
+
+        FlinkAssertions.assertThatChainOfCauses(ErrorHandlingSubmissionJob.getSubmissionException())
+                .as(
+                        "The job's main method shouldn't have been succeeded due to a DuplicateJobSubmissionException.")
+                .hasAtLeastOneElementOfType(DuplicateJobSubmissionException.class);
+
+        assertThat(
+                        jobResultStore.hasDirtyJobResultEntry(
+                                ApplicationDispatcherBootstrap.ZERO_JOB_ID))
+                .isTrue();
+        assertThat(
+                        jobResultStore.hasCleanJobResultEntry(
+                                ApplicationDispatcherBootstrap.ZERO_JOB_ID))
+                .isFalse();
```
This test is here to check that no job was re-submitted in case of a dirty jobresult existing in the `JobResultStore`. The cleanup is integrated in [FLINK-25432](https://issues.apache.org/jira/browse/FLINK-25432). That's where we would have to adjust the test accordingly.

We won't miss the change in `FLINK-25432` because the test would fail after integrating the cleanup because of the asserts on `hasDirtyJobResultEntry` and `hasCleanJobResultEntry`

dmvk reviewed (approved):
👍 Thanks for fixing the commit history. LGTM, let's follow-up with other PRs

XComp commented:
Thanks @dmvk . I rebased the branch once more to resolve some conflict caused by FLINK-24038 being merged. Additionally, I fixed a minor typo in a variable and updated the commit messages. I will merge the PR as soon as the final AzureCI run succeeds.

XComp commented:
The AzureCI build was successful. Merging now.
